### PR TITLE
Add GL90 diagnostics

### PR DESCRIPTION
**Pull request by NoraLoose**
This PR is a follow-up from PR https://github.com/NOAA-GFDL/MOM6/pull/268, and adds new diagnostics associated with the GL90 parameterization. An overview of the new diagnostics, alongside with examples, is given below.

I tried to keep a clean commit history (one commit per new diagnostic group, on December 8 and 13). However, I started from branch `NoraLoose:gl90-param` a couple of weeks ago, where the latter branch got squash-merged into `dev/gfdl` this morning (https://github.com/NOAA-GFDL/MOM6/pull/268). So things are not very clean anymore.

### Coupling coefficients and viscosities

* `au_gl90_visc`: zonal viscous coupling coefficient associated with GL90, is contained in `au_visc`
* `av_gl90_visc`: meridional viscous coupling coefficient associated with GL90, is contained in `av_visc`
* `Kv_gl90_u`: GL90 vertical viscosity at u-points, is contained in `Kv_u`
* `Kv_gl90_v`: GL90 vertical viscosity at v-points, is contained in `Kv_v`

**Example**:

av_and_avgl90_GL800
Kv_and_Kvgl90_GL800

### Momentum budget diagnostics

* `du_dt_visc_gl90`: zonal acceleration due to GL90 vertical viscosity, included in `du_dt_visc`
* `dv_dt_visc_gl90`: meridional acceleration due to GL90 vertical viscosity, included in `dv_dt_visc`

**Example**:

dv_dt_visc_and_dv_dt_visc_gl90_GL800

### Energy budget diagnostics

* GLwork: Kinetic Energy Source from GL90 Vertical Viscosity

**Example**:

The energetics of the GL90 parameterization (named `GLwork`) are intentionally computed in `MOM_vert_friction`, rather than in `MOM_diagnostics`, where the remaining kinetic energy budget terms are computed. We have to do the computation in `MOM_vert_friction` to ensure sign-definiteness when GLwork is summed in the vertical. Indeed, `MOM_diagnostics` does not have access to the velocities and thicknesses used in the vertical solver, but rather uses a time-mean barotropic transport `[uv]h` to compute the energy budget diagnostics. A detailed discussion and exploration of this issue can be found in https://github.com/ocean-eddy-cpt/MOM6/issues/25.

As a result of not computing the energetics in `MOM_diagnostics`, `GLwork` is not exactly contained in `KE_visc`. `KE_visc` represents the energetics of all vertical viscosity contributions, including the GL90 vertical viscosity. We could implement a term `KE_visc_gl90` that can be 1-to-1 compared to `KE_visc`; that is, `KE_visc` - `KE_visc_gl90` would represent exactly the energetics of all viscosity contributions EXCEPT the GL90 viscosity. If we implemented `KE_visc_gl90`, this term would in practice be very similar as `GLwork`, but sign-definiteness is not ensured, see above.

### CODE REVIEW

**Review by Hallberg-NOAA — APPROVED** [review]
Once again, the documentation of this pull request and the rationale behind it is absolutely superb.  Thank you for this valuable contribution, @NoraLoose!

  There are two issues here that I think are worth revisiting.

  1. I am uncomfortable with the idea of adding terms to the diagnostic energy budget that do not add up, so the proposed idea (aired in the discussion) of adding the alternative diagnostic of the contributions of GL to the energy budget, as calculated with the transports used in the other terms (even though GL is no longer sign definite) would seem to be a good idea.  This could be a reasonable follow-up to this PR, and it is not a reason not to accept this PR.

  2. The other thing that makes me a bit uneasy is that this PR is requiring that we add halo updates to this module.  In other places (for example in the lateral parameterization directory), the way that we have avoided this is by doing all of the calculations on a symmetric grid, even if we are using non-symmetric memory.  I think that we could probably do the same thing here, but doing so would go well beyond what would be reasonable to ask of this PR.

  So I am approving this PR as written (pending it passing the pipeline testing, of course), but I think that we should add also an issue to explore  whether or when it would make sense to to symmetric calculations for the vertical viscosity. I also think that the we should add the complementary (non-sign-definite) GL work diagnostics that could be used with the rest of the closed energy budget.

**Comment by NoraLoose** [comment]
Thanks for your review and comments, @Hallberg-NOAA! 

Re 1.: This would be really quick for me to do (I even have this implemented in another branch), so I will open another PR with the complementary GL work diagnostic.

Re 2.: I would probably need some help with this one. Do you mind opening the issue? I would have to dive a bit more into non-symmetric / symmetric memory handling before being able to state the problem clearly.

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17903 :heavy_check_mark: :yellow_circle: 

This has rebase conflicts but I will squash-merge it if OK with @NoraLoose.  (Two of the commits look significant, the others can probably be squashed).

**Comment by Hallberg-NOAA** [comment]
Because there were other commits merged into dev/gfdl after this PR was approved, than then had to be merged back onto this PR, I triggered the pipeline testing anew at https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17922.  Unfortunately, this is now failing at run-time with the gnu and pgi compilers, but not intel nor for the gnu-restarts runs, for reasons that are not clear.  It is possible that these failures have to do with the status of Gaea and its filesystems, rather than anything to do with this commit.

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/17907 :heavy_check_mark:  :yellow_circle: 

I will squash-merge this one.

**Pull request merged**
